opendevin/llm/llm.py

```diff
@@ -86,49 +86,28 @@ def __init__(
         """
         if llm_config is None:
             llm_config = config.llm
-        model = model if model is not None else llm_config.model
-        api_key = api_key if api_key is not None else llm_config.api_key
-        base_url = base_url if base_url is not None else llm_config.base_url
-        api_version = api_version if api_version is not None else llm_config.api_version
+        self.model_name = model or llm_config.model
+        self.api_key = api_key or llm_config.api_key
+        self.base_url = base_url or llm_config.base_url
+        self.api_version = api_version or llm_config.api_version
         num_retries = num_retries if num_retries is not None else llm_config.num_retries
         retry_min_wait = (
             retry_min_wait if retry_min_wait is not None else llm_config.retry_min_wait
         )
         retry_max_wait = (
             retry_max_wait if retry_max_wait is not None else llm_config.retry_max_wait
         )
-        llm_timeout = llm_timeout if llm_timeout is not None else llm_config.timeout
-        llm_temperature = (
-            llm_temperature if llm_temperature is not None else llm_config.temperature
-        )
+        self.llm_timeout = llm_timeout or llm_config.timeout
+        llm_temperature = llm_temperature or llm_config.temperature
         llm_top_p = llm_top_p if llm_top_p is not None else llm_config.top_p
-        custom_llm_provider = (
-            custom_llm_provider
-            if custom_llm_provider is not None
-            else llm_config.custom_llm_provider
-        )
-        max_input_tokens = (
-            max_input_tokens
-            if max_input_tokens is not None
-            else llm_config.max_input_tokens
-        )
-        max_output_tokens = (
-            max_output_tokens
-            if max_output_tokens is not None
-            else llm_config.max_output_tokens
-        )
-        metrics = metrics if metrics is not None else Metrics()
+        # not used in the current version
+        self.custom_llm_provider = custom_llm_provider or llm_config.custom_llm_provider
+        self.max_input_tokens = max_input_tokens or llm_config.max_input_tokens
+        self.max_output_tokens = max_output_tokens or llm_config.max_output_tokens
```

Comment:
Ditto, I don't know if 0 has meaning here, or will it have, but they're numeric params. So I'd assume if the user set 0, they reasonably expect us to send 0.

Reply:
If they pass `None`?
https://github.com/OpenDevin/OpenDevin/blob/6c741372ffff71a98f2f03bcc0dc4841cd444871/opendevin/llm/llm.py#L131-L137

Reply:
> If they pass `None`?
> 
> https://github.com/OpenDevin/OpenDevin/blob/6c741372ffff71a98f2f03bcc0dc4841cd444871/opendevin/llm/llm.py#L131-L137

`max_input_tokens` and `max_output_tokens` are already dealt with at that point:
https://github.com/OpenDevin/OpenDevin/blob/90ec0095dfe72e30114e093f2dc27c6bc75e8140/opendevin/llm/llm.py#L110-L119

Reply:
If the user set `None`, they also reasonably expect us to send `None`.

Reply:
> If the user set `None`, they also reasonably expect us to send `None`.

I'm not clear what you propose to change here?

_If_ the model returns actual values for min/max tokens, those values are hard facts and imho have precedence over any user-made config/param, even if None (talking about these "if self.min/max_output_tokens is None:" blocks).

Reply:
In a cursory search in the codebase I couldn't find that e.g. `max_tokens` or the min/max_output_tokens are actually used yet for other purposes than in the config.
Maybe these code locations should be revisited once that is the case?

Reply:
Currently, if the user gives None, the value is changed. 
0 is not a valid value.

Reply:
It is used here.
https://github.com/OpenDevin/OpenDevin/blob/6c741372ffff71a98f2f03bcc0dc4841cd444871/opendevin/llm/llm.py#L146

Reply:
Why should one set it to zero?

Reply:
Just to reset the context of what we are discussing here in class `LLM`:
the variables that are checked here are values coming from the codebase, e.g. if an agent would set a specific value - for whatever reason. Thus if they are not None, they are used and not the value coming from the `llm_config`. The code would rely the LLM class to obey this to make the deterministic and not infuse an unknown value from the _user_.
When saying _user_ it is what is set in the `config.toml`'s `llm` section, in `llm_config`.
The code here is mostly to determine **precedence** of values, runtime parameter vs. config.
And the `or` operator introduces behavioral change with regard to `falsey` values like 0 or empty string.
We are _not_ checking each and every value for validity coming from the `llm` section.
The _user_ of OpenDevin is responsible to configure sensible values.

Reply:
That latest commit is certainly an improvement, yes, but doesn't change my post above.